README.md
# ETTextField

_Last revision: 29.7.2020_

ETTextField is an extension of traditional UITextField which simplifies its customization and event handling. To get a quick idea jump to example section below!

![Demo](Docs/demo.gif)

## Features
- Fully customizable UI (colors, borders, insets, etc.)
- Floating title when user starts typing
- Error messages below text field with custom tint color
- Observation of user interaction with ETBinding events

## Usage `ETTextField`
### Create style for your text field with `TextFieldStyle`:
```swift
let style = TextFieldStyle(
    background: UIColor(white: 245.0/255.0, alpha: 1),
    font: UIFont.systemFont(ofSize: 14, weight: .light),
    tintColor: .blue,
    cornerRadius: 5.0,
    insets: UIEdgeInsets(top: 0, left: 10, bottom: 0, right: 10),
    border: [.bottom],
    borderColor: .blue
)
```
### Setup text field:
```swift
let textField = ETTextField()
textField.update(with: style)
textField.placeholder = "Enter your first name"
textField.title = "YOUR FIRST NAME:"
textField.titleOffset = -3.0
view.addSubview(textField)
```
### Observe user interaction:
```swift
textField.onDidBeginEditing.observe(owner: label) {
    // When user starts editing
}
textField.onDidEndEditing.observe(owner: label) {
    // When user ends editing
}
textField.onDidChangeText.observe(owner: label) { text in
    // When input text changes
}
textField.onReturnKeyPressed.observe(owner: self) {
    // When user hits return
}
```

### Error messages:
#### Show
```swift
textField.showError(message: "Error example")
```

#### Hide
The error message is automatically hidden if any change occurs on the `textField`. 

You can explicitly hide the error message:
```swift
textField.hideError()
```

## Usage `ETCustomErrorTextField`

### Setup text field:
The difference from `ETTextField` is that you can inject custom view and optionaly custom error icon that will be displayed on the right side of the field.
There's also the possibility to show custom error view only while user edits the textfield. For this you need to call `showError(onFocusOnly: true)`.
```swift
let textField = ETCustomErrorTextField(errorView: self.makeCustomErrorView())
```
or with custom error icon
```swift
let textField = ETCustomErrorTextField(errorView: self.makeCustomErrorView(), errorIcon: UIImage(named: "ic_error_icon"))
```
You can also inherit from the class.

For more info see included example project.

## Installation

### Carthage

Add `github "EtneteraMobile/ETTextField"` to your Cartfile.

### Swift Package Manager

In Xcode (>11.0) go to File -> Swift Packages -> Add Package Dependency. There insert `https://github.com/EtneteraMobile/ETTextField` in URL input and finish importing `ETTextField` to your project.

## Contributing

Contributions to ETTextField are welcomed and encouraged!

## License

ETTextField is available under the MIT license. See [LICENSE](LICENSE) for more information.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.0.5] - 2021-04-14

### Fixed

- Fixed `Package.swift` for support minimal swift version 4.2 and fix config of dependency

## [1.0.4] - 2020-11-24

### Added

- Added support for custom error color.

## [1.0.3] - 2020-08-05

### Added

- `ETCustomErrorTextField` is now open class and you can inherit from it.

## [1.0.2] - 2020-08-03

### Fixed
- Fixed  `ETCustomErrorTextField`'s overided method `textFieldDidEndEditing` calling `textFieldDidBeginEditing` in it's parent. 

## [1.0.1] - 2020-07-29

### Added

- Added custom error icon to show in ETCustomErrorTextField
- Added posibility to show custom error view only while editing the text field. 

## [1.0.0] - 2020-07-14

### Added

- New textfield variant with custom error view
- Swiftlint build phase

## [0.1.5] - 2020-06-15

### Added

- Title background color support, disabled tint color
- Added README and CHANGELOG


Example/ExampleController.swift
//
//  ExampleController.swift
//  Example
//
//  Created by Beles on 10/08/2019.
//  Copyright © 2019 ETTextField. All rights reserved.
//

import ETTextField
import UIKit

class ExampleController: UIViewController {

    private let firstNameTF = ETTextField()
    private let secondNameTF = ETTextField()
    private lazy var customErrorTF = ETCustomErrorTextField(errorView: self.makeCustomErrorView())
    private lazy var customErrorWithIconTF = ETCustomErrorTextField(errorView: self.makeCustomErrorView(), errorIcon: UIImage(named: "ic_error_icon"))
    private let button = UIButton()

    override func viewDidLoad() {
        super.viewDidLoad()
        view.backgroundColor = .white
        setupFirstNameTF()
        setupSecondNameTF()
        setupCustomErrorTF()
        setupCustomErrorWithIconTF()
        setupShowErrorButton()
        setupObservers()

        let tapRecognizer = UITapGestureRecognizer(target: self, action: #selector(didTapOutside))
        view.addGestureRecognizer(tapRecognizer)
    }

    private func setupFirstNameTF() {
        let style = TextFieldStyle(background: #colorLiteral(red: 0.9499999881, green: 0.9499999881, blue: 0.9499999881, alpha: 1) ,
                                   font: UIFont.systemFont(ofSize: 14, weight: .light),
                                   tintColor: .blue,
                                   cornerRadius: 5.0,
                                   insets: UIEdgeInsets(top: 0, left: 10, bottom: 0, right: 10),
                                   border: [.bottom],
                                   borderColor: .blue)
        firstNameTF.update(with: style)
        firstNameTF.placeholder = "Enter your first name"
        firstNameTF.title = "YOUR FIRST NAME:"
        firstNameTF.titleOffset = -3.0

        view.addSubview(firstNameTF)
        firstNameTF.translatesAutoresizingMaskIntoConstraints = false

        firstNameTF.centerXAnchor.constraint(equalTo: view.centerXAnchor).isActive = true
        firstNameTF.topAnchor.constraint(equalTo: view.safeAreaLayoutGuide.topAnchor, constant: 50).isActive = true
        firstNameTF.widthAnchor.constraint(equalTo: view.widthAnchor, multiplier: 0.8).isActive = true
        firstNameTF.heightAnchor.constraint(equalToConstant: 30).isActive = true
    }

    private func setupSecondNameTF() {
        let style = TextFieldStyle(background: .clear,
                                   font: UIFont.systemFont(ofSize: 14, weight: .light),
                                   tintColor: .blue,
                                   cornerRadius: 0.0,
                                   insets: UIEdgeInsets(top: 0, left: 10, bottom: 0, right: 10),
                                   border: [.top, .right, .bottom, .left],
                                   borderColor: .lightGray)
        secondNameTF.update(with: style)
        secondNameTF.placeholder = "Enter your last name"
        secondNameTF.title = "YOUR LAST NAME:"
        secondNameTF.titleOffset = -3.0

        view.addSubview(secondNameTF)
        secondNameTF.translatesAutoresizingMaskIntoConstraints = false

        secondNameTF.centerXAnchor.constraint(equalTo: view.centerXAnchor).isActive = true
        secondNameTF.topAnchor.constraint(equalTo: firstNameTF.bottomAnchor, constant: 45).isActive = true
        secondNameTF.widthAnchor.constraint(equalTo: view.widthAnchor, multiplier: 0.8).isActive = true
        secondNameTF.heightAnchor.constraint(equalToConstant: 30).isActive = true
    }

    private func setupCustomErrorTF() {
        let style = TextFieldStyle(background: .clear,
                                   font: UIFont.systemFont(ofSize: 14, weight: .light),
                                   tintColor: .blue,
                                   cornerRadius: 0.0,
                                   insets: UIEdgeInsets(top: 0, left: 10, bottom: 0, right: 10),
                                   borderColor: .lightGray)

        customErrorTF.update(with: style)
        customErrorTF.autocapitalizationType = .none
        customErrorTF.placeholder = "Enter your e-mail"
        customErrorTF.title = "YOUR E-MAIL"

        view.addSubview(customErrorTF)
        customErrorTF.translatesAutoresizingMaskIntoConstraints = false

        customErrorTF.centerXAnchor.constraint(equalTo: view.centerXAnchor).isActive = true
        customErrorTF.topAnchor.constraint(equalTo: secondNameTF.bottomAnchor, constant: 45).isActive = true
        customErrorTF.widthAnchor.constraint(equalTo: view.widthAnchor, multiplier: 0.8).isActive = true
    }

    private func setupCustomErrorWithIconTF() {
        let style = TextFieldStyle(background: .clear,
                                   font: UIFont.systemFont(ofSize: 14, weight: .light),
                                   tintColor: .blue,
                                   cornerRadius: 0.0,
                                   insets: UIEdgeInsets(top: 0, left: 10, bottom: 0, right: 10),
                                   borderColor: .lightGray)

        customErrorWithIconTF.update(with: style)
        customErrorWithIconTF.autocapitalizationType = .none
        customErrorWithIconTF.placeholder = "Repeat your e-mail"
        customErrorWithIconTF.title = "YOUR E-MAIL"

        view.addSubview(customErrorWithIconTF)
        customErrorWithIconTF.translatesAutoresizingMaskIntoConstraints = false

        customErrorWithIconTF.centerXAnchor.constraint(equalTo: view.centerXAnchor).isActive = true
        customErrorWithIconTF.topAnchor.constraint(equalTo: customErrorTF.bottomAnchor, constant: 45).isActive = true
        customErrorWithIconTF.widthAnchor.constraint(equalTo: view.widthAnchor, multiplier: 0.8).isActive = true
    }

    private func setupShowErrorButton() {
        button.setTitle("Show errors", for: .normal)
        button.setTitleColor(.white, for: .normal)
        button.backgroundColor = .blue
        view.addSubview(button)
        button.translatesAutoresizingMaskIntoConstraints = false
        button.centerXAnchor.constraint(equalTo: view.centerXAnchor).isActive = true
        button.topAnchor.constraint(equalTo: customErrorWithIconTF.bottomAnchor, constant: 35).isActive = true
        button.widthAnchor.constraint(equalTo: view.widthAnchor, multiplier: 0.8).isActive = true
        button.heightAnchor.constraint(equalToConstant: 30).isActive = true
        button.addTarget(self, action: #selector(showError), for: .touchUpInside)
    }

    private func setupObservers() {
        let label = UILabel()
        label.textAlignment = .center
        label.numberOfLines = 0
        view.addSubview(label)
        label.translatesAutoresizingMaskIntoConstraints = false
        label.centerXAnchor.constraint(equalTo: view.centerXAnchor).isActive = true
        label.topAnchor.constraint(equalTo: button.bottomAnchor, constant: 35).isActive = true
        label.widthAnchor.constraint(equalTo: view.widthAnchor).isActive = true

        firstNameTF.onDidBeginEditing.observe(owner: label) {
            label.text = "firstNameTF: Did begin editing"
        }

        firstNameTF.onDidEndEditing.observe(owner: label) {
            label.text = "firstNameTF: Did end editing"
        }

        firstNameTF.onDidChangeText.observe(owner: label) { text in
            label.text = "firstNameTF: \(text ?? "<empty>")"
        }

        firstNameTF.onReturnKeyPressed.observe(owner: self) {
            label.text = "firstNameTF: Did press return key"
            self.secondNameTF.becomeFirstResponder()
        }

        secondNameTF.onDidBeginEditing.observe(owner: label) {
            label.text = "secondNameTF: Did begin editing"
        }

        secondNameTF.onDidEndEditing.observe(owner: label) {
            label.text = "secondNameTF: Did end editing"
        }

        secondNameTF.onDidChangeText.observe(owner: label) { text in
            label.text = "secondNameTF: \(text ?? "<empty>")"
        }

        secondNameTF.onReturnKeyPressed.observe(owner: label) {
            label.text = "secondNameTF: Did press return key"
        }

        customErrorTF.onDidBeginEditing.observe(owner: label) {
            label.text = "customErrorTF: Did begin editing"
        }

        customErrorTF.onDidEndEditing.observe(owner: label) {
            label.text = "customErrorTF: Did end editing"
        }

        customErrorTF.onDidChangeText.observe(owner: label) { text in
            label.text = "customErrorTF: \(text ?? "<empty>")"
        }

        customErrorTF.onReturnKeyPressed.observe(owner: label) {
            label.text = "customErrorTF: Did press return key"
        }

        customErrorWithIconTF.onDidBeginEditing.observe(owner: label) {
            label.text = "customErrorWithIconTF: Did begin editing"
        }

        customErrorWithIconTF.onDidEndEditing.observe(owner: label) {
            label.text = "customErrorWithIconTF: Did end editing"
        }

        customErrorWithIconTF.onDidChangeText.observe(owner: label) { text in
            label.text = "customErrorWithIconTF: \(text ?? "<empty>")"
        }

        customErrorWithIconTF.onReturnKeyPressed.observe(owner: label) {
            label.text = "customErrorWithIconTF: Did press return key"
        }
    }

    private func makeCustomErrorView() -> UIView {
        let customErrorView = UIView()
        customErrorView.backgroundColor = .red

        let errorLabel = UILabel()
        errorLabel.numberOfLines = 0
        errorLabel.lineBreakMode = .byWordWrapping
        errorLabel.text = "Wrong e-mail format"
        errorLabel.textColor = .white
        errorLabel.font = UIFont.systemFont(ofSize: 12)
        errorLabel.translatesAutoresizingMaskIntoConstraints = false

        customErrorView.addSubview(errorLabel)

        NSLayoutConstraint.activate([
            errorLabel.topAnchor.constraint(equalTo: customErrorView.topAnchor, constant: 5),
            errorLabel.leadingAnchor.constraint(equalTo: customErrorView.leadingAnchor, constant: 5),
            errorLabel.trailingAnchor.constraint(equalTo: customErrorView.trailingAnchor, constant: -5),
            errorLabel.bottomAnchor.constraint(equalTo: customErrorView.bottomAnchor, constant: -5)
        ])
        return customErrorView
    }

    @objc
    private func showError() {
        firstNameTF.showError(message: "Error example")
        secondNameTF.showError(message: "Error example")
        customErrorTF.showError()
        customErrorWithIconTF.showError(onFocusOnly: true)
    }

    @objc
    private func didTapOutside() {
        view.endEditing(true)
    }
}
